lib/charms/loki_k8s/v1/loki_push_api.py

```diff
@@ -2313,6 +2387,120 @@ def _containers(self) -> Dict[str, Container]:
+
+    def _on_logging_relation_changed(self, _):
+        self.enable()
+
+    def _on_logging_relation_departed(self, _):
+        self.disable()
+
+    def _on_logging_relation_broken(self, _):
+        self.disable()
+
+    def enable(self):
+        """Enable log forwarding."""
+        self.handle_logging(enabled=True)
+
+    def disable(self):
+        """Disable log forwarding."""
+        self.handle_logging(enabled=False)
+
+    def _build_log_target(self, endpoint, index, enabled=False):
+        """Build a log target for the log forwarding Pebble layer."""
+        dest_name = f"loki{index}"
+        services_value = ["all"] if enabled else ["-all"]
+
+        return {
+            dest_name: {
+                "override": "merge",
+                "type": "loki",
+                "location": endpoint,
+                "services": services_value,
+                "labels": {
+                    "product": "Juju",
+                    "charm": self.topology._charm_name,
+                    "juju_model": self.topology._model,
+                    "juju_model_uuid": self.topology._model_uuid,
+                    "juju_application": self.topology._application,
+                    "juju_unit": self.topology._unit,
+                },
+            }
+        }
+
+    def _build_log_targets(self, endpoints: Optional[List[str]], enable=False):
+        """Build the targets for the log forwarding Pebble layer."""
+        targets = {}
+        if endpoints:
+            for i, endpoint in enumerate(endpoints):
+                targets.update(self._build_log_target(endpoint, i, enable))
+        return targets
+
+    def handle_logging(self, enabled=False):
+        """Enable or disable the log forwarding."""
+        if self.loki_endpoints:
+            layer_config = {"log-targets": self._build_log_targets(self.loki_endpoints, enabled)}
+            layer = Layer(layer_config)  # pyright: ignore
+
+            for container_name, container in self._charm.unit.containers.items():
+                container.add_layer(f"{container_name}-log-forwarding", layer, combine=True)
+        else:
+            logger.warning("No loki endpoints available")
+
+    def _fetch_endpoints(self, relation_name: str) -> List[str]:
+        """Fetch Loki Push API endpoints through relation data.
+
+        Returns:
+            A list of Loki Push API endpoint URLs, for instance:
+            [
+                "http://loki1:3100/loki/api/v1/push",
+                "http://loki2:3100/loki/api/v1/push",
+            ]
+        """
+        endpoints = []  # type: list
+
+        for relation in self._charm.model.relations[relation_name]:
+            for unit in relation.units:
+                if unit.app == self._charm.app:
+                    # This is a peer unit
+                    continue
+
+                endpoint = relation.data[unit].get("endpoint")
+                if endpoint:
+                    deserialized_endpoint = json.loads(endpoint)
```

Comment:
2) if this fails, do we just raise an uncaught error? This will bork the charm's initialization